src/core/cgroup-bpf-ctx.c

```diff
@@ -0,0 +1,58 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+#include "cgroup-bpf-ctx.h"
+#include "cgroup.h"
+#include "memory-util.h"
+#include "strv.h"
+#include "string-util.h"
+
+int cgroup_add_pinned_bpf_program(CGroupContext *c, enum bpf_attach_type attach_type, const char *bpffs_path) {
+        _cleanup_free_ CGroupPinnedBPFProgram *p = NULL;
+        _cleanup_free_ char *dup = NULL;
+
+        assert(c);
+        assert(bpffs_path);
+
+        if (cgroup_contains_pinned_bpf_program(c, attach_type, bpffs_path))
+                return log_warning_errno(SYNTHETIC_ERRNO(EEXIST),
+                                "BPF program %s is already specified in cgroup context", bpffs_path);
+
+        dup = strdup(bpffs_path);
+        if (!dup)
+                return log_oom();
+
+        p = new(CGroupPinnedBPFProgram, 1);
+        if (!p)
+                return log_oom();
+
+        *p = (CGroupPinnedBPFProgram) {
+                .attach_type = attach_type,
+                .bpffs_path = TAKE_PTR(dup),
+        };
+
+        LIST_PREPEND(prog, c->pinned_bpf_progs, TAKE_PTR(p));
+
+        return 0;
+}
+
+bool cgroup_contains_pinned_bpf_program(
+                CGroupContext *c, enum bpf_attach_type attach_type, const char *bpffs_path) {
+        CGroupPinnedBPFProgram *p, *n;
+
+        assert(c);
+        assert(bpffs_path);
+
+        LIST_FOREACH_SAFE(prog, p, n, c->pinned_bpf_progs)
+                if (p->attach_type == attach_type && streq(p->bpffs_path, bpffs_path))
```

Comment:
file system paths should be compare with path_equal()

Reply:
@poettering 

This comparison is also the reason why it's necessary to store paths in `CGroupContext`, unless we are fine with attaching the same program more than once.

Reply:
hmm, but an inode nr check would be better for that, no?

Reply:
@poettering 

Addressed comments, please take a look.
Checking for program uniqueness by BPF program ID. Same IDs essentially means that is the same program, i.e. the chunk of occupied kernel memory is the same, FDs or paths may be different. The idea is somewhat similar to inode describing filesystem object.
Introduced `bpf_program_get_id_by_fd` helper for that. 
Program ID and attach type form a lookup key for "foreign" BPF program, so the same program is attached only once for each attach point specified.